File: Tools/Sources/StuffToolCore/DeviceSelection.swift
import Foundation

public struct DeviceSelection: Equatable, Sendable {
    public let identifier: String
    public let name: String
    public let connectionState: String

    public init(identifier: String, name: String, connectionState: String) {
        self.identifier = identifier
        self.name = name
        self.connectionState = connectionState
    }
}

public enum DeviceSelectionFailure: Error, Equatable, CustomStringConvertible, Sendable {
    case message(String)

    public var description: String {
        switch self {
            case let .message(message): message
        }
    }
}

/// Selects one physical iOS-family device from `devicectl`'s typed JSON schema.
public struct DeviceSelector: Sendable {
    public init() {}

    public func select(from data: Data, filter: String?) throws -> DeviceSelection {
        let list: DeviceControlList
        do {
            list = try JSONDecoder().decode(DeviceControlList.self, from: data)
        } catch {
            throw DeviceSelectionFailure.message(
                "couldn't read devicectl's device list (\(error)). " +
                    "Run `xcrun devicectl list devices` to check your setup.",
            )
        }
        let wanted = filter?.trimmingCharacters(in: .whitespacesAndNewlines).lowercased() ?? ""
        let candidates = (list.result?.devices ?? []).compactMap { device -> DeviceSelection? in
            let hardware = device.properties?.hardware
            guard ["ios", "ipados"].contains(hardware?.platform?.lowercased() ?? ""),
                  hardware?.reality?.lowercased() == "physical"
            else {
                return nil
            }
            let identifier = nonempty(device.identifier) ?? nonempty(hardware?.udid)
            guard let identifier else { return nil }
            let name = nonempty(device.properties?.state?.name) ?? "(unnamed)"
            let udid = hardware?.udid ?? ""
            if wanted.isEmpty == false,
               [name.lowercased(), udid.lowercased(), identifier.lowercased()]
               .contains(wanted) == false
            {
                return nil
            }
            return DeviceSelection(
                identifier: identifier,
                name: name,
                connectionState: nonempty(device.properties?.connection?.state) ?? "unknown",
            )
        }

        guard candidates.isEmpty == false else {
            if let filter, wanted.isEmpty == false {
                throw DeviceSelectionFailure.message(
                    "no physical iOS device matching \"\(filter)\". " +
                        "Run `xcrun devicectl list devices` to see what's paired.",
                )
            }
            throw DeviceSelectionFailure.message(
                "no physical iOS device found. Pair or plug in an iPhone, unlock it, " +
                    "and trust this Mac.",
            )
        }
        guard candidates.count == 1 else {
            let listing = candidates.map {
                "  - \($0.name) (\($0.identifier)) [\($0.connectionState)]"
            }.joined(separator: "\n")
            throw DeviceSelectionFailure.message(
                "multiple physical iOS devices; pass --device <name|udid>:\n\(listing)",
            )
        }
        return candidates[0]
    }

    private func nonempty(_ value: String?) -> String? {
        guard let value, value.isEmpty == false else { return nil }
        return value
    }
}

private struct DeviceControlList: Decodable {
    let result: Result?

    struct Result: Decodable {
        let devices: [Device]?
    }

    struct Device: Decodable {
        let identifier: String?
        let properties: Properties?
    }

    struct Properties: Decodable {
        let hardware: Hardware?
        let connection: Connection?
        let state: State?
    }

    struct Hardware: Decodable {
        let platform: String?
        let reality: String?
        let udid: String?
    }

    struct Connection: Decodable {
        let state: String?
    }

    struct State: Decodable {
        let name: String?
    }
}

File: Tools/device_selection.rb
# frozen_string_literal: true

require "json"
require "optparse"

# Selects one exact physical iOS device from devicectl's JSON output.
class DeviceSelection
  class Error < StandardError; end

  Selection = Struct.new(:identifier, :name, :connection, keyword_init: true)

  def self.select(json:, filter:)
    data = JSON.parse(json)
    devices = data.dig("result", "devices")
    raise Error, "devicectl's result has no devices array" unless devices.is_a?(Array)

    wanted = filter.strip.downcase
    candidates = devices.filter_map do |device|
      next unless device.is_a?(Hash)

      properties = device["properties"]
      next unless properties.is_a?(Hash)

      hardware = properties["hardware"] || {}
      connection = properties["connection"] || {}
      state = properties["state"] || {}
      next unless %w[ios ipados].include?(hardware.fetch("platform", "").to_s.downcase)
      next unless hardware.fetch("reality", "").to_s.downcase == "physical"

      identifier = device["identifier"] || hardware["udid"]
      next unless identifier.is_a?(String) && !identifier.empty?

      name = state["name"].is_a?(String) && !state["name"].empty? ? state["name"] : "(unnamed)"
      udid = hardware["udid"].to_s
      next if !wanted.empty? && ![name, udid, identifier].any? { |value| value.downcase == wanted }

      Selection.new(
        identifier: identifier,
        name: name,
        connection: connection["state"].to_s.empty? ? "unknown" : connection["state"],
      )
    end

    if candidates.empty?
      if wanted.empty?
        raise Error, "no physical iOS device found. Pair or plug in an iPhone, unlock it, and trust this Mac."
      end
      raise Error, %(no physical iOS device matching "#{filter}". Run `xcrun devicectl list devices` to see what's paired.)
    end
    if candidates.length > 1
      listing = candidates.map { |candidate| "  - #{candidate.name} (#{candidate.identifier}) [#{candidate.connection}]" }.join("\n")
      raise Error, "multiple physical iOS devices; pass --device <name|udid>:\n#{listing}"
    end

    candidates.fetch(0)
  rescue JSON::ParserError => error
    raise Error, "couldn't read devicectl's device list (#{error.message}). Run `xcrun devicectl list devices` to check your setup."
  end
end

def device_selection_main(arguments)
  filter = ""
  parser = OptionParser.new do |options|
    options.on("--filter VALUE") { |value| filter = value }
  end
  parser.parse!(arguments)
  selection = DeviceSelection.select(json: $stdin.read, filter: filter)
  warn "    using: #{selection.name} (#{selection.identifier})"
  $stdout.write("#{selection.identifier}\t#{selection.name}")
  0
rescue DeviceSelection::Error, OptionParser::ParseError => error
  warn "error: #{error.message}"
  1
end

exit(device_selection_main(ARGV)) if __FILE__ == $PROGRAM_NAME
